File: Palestra di Dennis Ritchie (IMP)/Trading System/Trading.java
package BattleAndTrading;

import java.util.Scanner;

class Trainer {
    String name;
    String[] pokemonInventory;
    int pokemonCount;

    public Trainer(String name) {
        this.name = name;
        this.pokemonInventory = new String[10];
        this.pokemonCount = 0;
    }

    public void addPokemon(String pokemon) {
        if (pokemonCount < pokemonInventory.length) {
            pokemonInventory[pokemonCount++] = pokemon;
            System.out.println(pokemon + " added to " + name + "'s inventory.");
        } else {
            System.out.println(name + "'s inventory is full!");
        }
    }

    public void removePokemon(String pokemon) {
        for (int i = 0; i < pokemonCount; i++) {
            if (pokemonInventory[i].equals(pokemon)) {
                for (int j = i; j < pokemonCount - 1; j++) {
                    pokemonInventory[j] = pokemonInventory[j + 1];
                }
                pokemonInventory[--pokemonCount] = null;
                System.out.println(pokemon + " removed from " + name + "'s inventory.");
                return;
            }
        }
        System.out.println(pokemon + " not found in " + name + "'s inventory.");
    }

    public void displayInventory() {
        System.out.println(name + "'s Inventory:");
        if (pokemonCount == 0) {
            System.out.println("No Pokémon in inventory.");
        } else {
            for (int i = 0; i < pokemonCount; i++) {
                System.out.println(pokemonInventory[i]);
            }
        }
    }

    public boolean hasPokemon(String pokemon) {
        for (int i = 0; i < pokemonCount; i++) {
            if (pokemonInventory[i].equals(pokemon)) {
                return true;
            }
        }
        return false;
    }
}

public class Trading {

    static Trainer[] trainers = new Trainer[10];
    static int trainerCount = 0;

    public static Trainer findTrainer(String name) {
        for (int i = 0; i < trainerCount; i++) {
            if (trainers[i].name.equals(name)) {
                return trainers[i];
            }
        }
        return null;
    }

    public static Trainer addTrainer(String name) {
        Trainer existingTrainer = findTrainer(name);
        if (existingTrainer != null) {
            return existingTrainer;
        }

        if (trainerCount < trainers.length) {
            Trainer newTrainer = new Trainer(name);
            trainers[trainerCount++] = newTrainer;
            return newTrainer;
        } else {
            System.out.println("Max number of trainers reached!");
            return null;
        }
    }

    public static void tradePokemon(String trainerFrom, String trainerTo, String pokemon) {
        Trainer fromTrainer = findTrainer(trainerFrom);
        Trainer toTrainer = findTrainer(trainerTo);

        if (fromTrainer == null || toTrainer == null) {
            System.out.println("One or both trainers not found.");
            return;
        }

        if (fromTrainer.hasPokemon(pokemon)) {
            fromTrainer.removePokemon(pokemon);
            toTrainer.addPokemon(pokemon);
            System.out.println(trainerFrom + " traded " + pokemon + " to " + trainerTo + ".");
        } else {
            System.out.println(pokemon + " not found in " + trainerFrom + "'s inventory.");
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("\nOptions: [1] Add Pokémon [2] Remove Pokémon [3] Trade Pokémon [4] Show Inventory [5] Quit");
            System.out.print("Choose an option: ");
            String choice = scanner.nextLine();

            switch (choice) {
                case "1":
                    System.out.print("Enter trainer name: ");
                    String trainerAdd = scanner.nextLine();
                    System.out.print("Enter Pokémon to add: ");
                    String pokemonAdd = scanner.nextLine();
                    Trainer trainer1 = addTrainer(trainerAdd);
                    if (trainer1 != null) {
                        trainer1.addPokemon(pokemonAdd);
                    }
                    break;

                case "2":
                    System.out.print("Enter trainer name: ");
                    String trainerRemove = scanner.nextLine();
                    System.out.print("Enter Pokémon to remove: ");
                    String pokemonRemove = scanner.nextLine();
                    Trainer trainer2 = findTrainer(trainerRemove);
                    if (trainer2 != null) {
                        trainer2.removePokemon(pokemonRemove);
                    } else {
                        System.out.println(trainerRemove + " not found.");
                    }
                    break;

                case "3":
                    System.out.print("Enter trainer name to trade from: ");
                    String trainerFrom = scanner.nextLine();
                    System.out.print("Enter trainer name to trade to: ");
                    String trainerTo = scanner.nextLine();
                    System.out.print("Enter Pokémon to trade: ");
                    String pokemonTrade = scanner.nextLine();
                    tradePokemon(trainerFrom, trainerTo, pokemonTrade);
                    break;

                case "4":
                    System.out.print("Enter trainer name: ");
                    String trainerDisplay = scanner.nextLine();
                    Trainer trainer3 = findTrainer(trainerDisplay);
                    if (trainer3 != null) {
                        trainer3.displayInventory();
                    } else {
                        System.out.println(trainerDisplay + " not found.");
                    }
                    break;

                case "5":
                    System.out.println("Exiting the system.");
                    scanner.close();
                    return;

                default:
                    System.out.println("Invalid option. Please choose again.");
            }
        }
    }
}

File: Palestra di Dennis Ritchie (IMP)/Trading System/Team11/Trading.py
def menu():
    print("[1] Add Pokemon")
    print("[2] Remove Pokemon")
    print("[3] Trade Pokemon")
    print("[4] Show Inventory")
    print("[0] Quit")
    choice = int(input("Select a choice [0-4]: "))
    return choice

def add_pokemon(inventory, Trainer, pokemon):
    t =  get_trainer_by_name(Trainer, inventory)
    if(t == 0):
       new_trainer =  {'Name': Trainer, 'Pokemon': []}
       inventory.append(new_trainer)
       t =  get_trainer_by_name(Trainer, inventory)
    t['Pokemon'].append(pokemon)
    print(pokemon, " added to ", t['Name'], "'s inventory.")
      



def remove_pokemon(inventory, Trainer, pokemon):
    t =  get_trainer_by_name(Trainer, inventory)
    if( t != 0):
        if(pokemon in t['Pokemon']):
            t['Pokemon'].remove(pokemon)
            print(pokemon, " removed from "+ t['Name'],"'s inventory")
        else:
            print(pokemon, " not found in "+ t['Name'],"'s inventory")
    else:
        print(Trainer , " not found.")

def display_inventory(inventory, Trainer):
    t =  get_trainer_by_name(Trainer, inventory)
    if(t != 0):
        print(t['Name'], "'s Inventory")
        for p in t['Pokemon']:
            print(p, " ")
    else:
        print(Trainer , " not found.")

def trade_pokemon(inventory, trainer_from, trainer_to, pokemon):
    t1 = get_trainer_by_name(trainer_from, inventory)
    t2 = get_trainer_by_name(trainer_to, inventory)
    if(t1 != 0 and t2 != 0):
        if(pokemon in t1['Pokemon']):
            remove_pokemon(inventory,t1['Name'],pokemon)
            add_pokemon(inventory,t2['Name'],pokemon)
            print(t1['Name'], " traded ", pokemon, " to " , t2['Name'])
        else:
            print(pokemon, " not found in ", t1['Name'] , " inventory")
    else:
        print("One or both trainerso not found")

def get_trainer_by_name(name, inventory):
    for trainer in inventory:
        if trainer['Name'] == name:
            return trainer
    return 0

def switch_choice(inventory, sc):
     match sc:
        case 0:
            return
        case 1:
            t = input("Enter Trainer Name: ")
            p = input("Enter Pokemon to add: ")
            add_pokemon(inventory, t, p)
        case 2:
            t = input("Enter Trainer Name: ")
            p = input("Enter Pokemon to remove: ")
            remove_pokemon(inventory, t, p)
        case 3:
            t1 = input("Enter Trainer Name to trade from: ")
            t2 = input("Enter Trainer Name to trade to: ")
            p = input("Enter Pokemon to trade: ")
            trade_pokemon(inventory, t1, t2, p)
        case 4:
            t = input("Enter Trainer Name to display: ")
            display_inventory(inventory, t)
        case _:
            print("Invalid choice")

def main():
    Trainer1 = {'Name': 'Ash', 'Pokemon': ["Pikachu"]}
    Trainer2 = {'Name': 'Giovanni', 'Pokemon': ["Persian"]}
    inventory = [Trainer1, Trainer2]

    sc = ""
    while(sc != 0):
        sc = int(menu())
        switch_choice(inventory,sc)

    
main()